### ISSUE TITLE
How to spread existing object props along with Feliz props?

### ISSUE THREAD

**Issue by cmeeren** [Zaid-Ajaj/Feliz#38]
I'm using/looking at things like [downshift](https://github.com/downshift-js/downshift) and [downshift-hooks](https://github.com/silviuavram/downshift-hooks) which provide props that you're supposed to spread, for example:

```jsx
<label {...getLabelProps()}>Enter a fruit</label>
```

where `getLabelProps` is provided by the component's render function (downshift) or hook (downshift-hooks), and of course should be able to be combined with other custom props.

I'm also needing the same functionality for the Material-UI styling solution, which uses at least one mixin object in the theme (`theme.mixins.toolbar`) that you can combine with your own styles for a given class.

My current solution is to use this function to convert a JS object into key-value pairs:

```f#
[<Emit("Object.entries($0)")>]
/// Transforms a plain JS object to key-value pairs.
let objectEntries (x: 'a) : (string * obj) [] = jsNative
```

Then I have the following function which simply unboxes that to the needed type:

```f#
type prop =

  /// Converts the items in the object to an array of IReactProperty so they can
  /// be used alongside other props. Use yield! to combine these in a list with
  /// other props.
  static member spread(value: obj) : IReactProperty [] =
    objectEntries value |> unbox
```

(`style.spread` would be identical, except returning `IStyleAttribute []`.)

This allows usage like:

```f#
Html.div [
  yield! prop.spread (getLabelProps())
  yield prop.className "foo"
]
```

My question now is two-fold:

1. Are there better ways to accomplish this?
2. Since this is completely general and might be needed in many situations, would you consider adding support for this in Feliz?

**Comment by Zaid-Ajaj** [comment]
> Are there better ways to accomplish this?

Not that I know of, I am glad you already found a workaround

> Since this is completely general and might be needed in many situations, would you consider adding support for this in Feliz?

Actually I think this usage of `getLabelProps()` is too low-level for a Fable binding, the idea of a spread is a Javascript thing and when building Fable bindings, I believe that one should keep F#-y things only, so if you are building a binding for downshift, I would expect the binding to give me:
```fs
getLabelProps : unit -> IReactProperty list
```
which wraps the internal `getLabelProps` with  `objectEntries` in the binding implementation (i.e. `getLabelProps >> objectEntries`) then I could `yield!` with the other properties:
```fs
Html.div [
  yield! (getLabelProps())
  yield prop.className "foo"
]
```
I guess the same applies for spreading a style

**Comment by cmeeren** [comment]
I see your point, I agree in the context of `getLabelProps` (I have written bindings for downshift and know it's not hard to do there).

But what about MUI's mixin? It's defined some levels down (`theme.mixins.toolbar`) in interfaces that are only used for directly accessing native JS code, as is commonly seen in bindings. Changing this to return something else would mean that everything up the hierarchy would also need to be modified to not be interfaces but instead classes with custom logic. It seems like this can, in general, cause a potentially big maintenance step up from (slightly tweaked) ts2fable bindings, just to change this one thing. And this would be necessary for all projects where this is relevant.

In short, having this possibility in Feliz would make things a lot easier for maintainers of bindings where this is necessary.

I agree that it's low-level and that ideally, bindings returning something to be spread would return `IReactProperty list`, but it can cause notably higher maintenance.

Just sharing thoughts, not going to force this into Feliz of course. :)

**Comment by Zaid-Ajaj** [comment]
> But what about MUI's mixin? It's defined some levels down (theme.mixins.toolbar) in interfaces that are only used for directly accessing native JS code, as is commonly seen in bindings. Changing this to return something else would mean that everything up the hierarchy would also need to be modified to not be interfaces but instead classes with custom logic. It seems like this can, in general, cause a potentially big maintenance step up from (slightly tweaked) ts2fable bindings, just to change this one thing. And this would be necessary for all projects where this is relevant.

Can you show me an example? 

> In short, having this possibility in Feliz would make things a lot easier for maintainers of bindings where this is necessary.

I still think these primitives shouldn't be exposed through Feliz, nor the binding that depends on it. 

> I agree that it's low-level and that ideally, bindings returning something to be spread would return IReactProperty [], but it can cause notably higher maintenance.

IMO, correctness and API ergonomics > maintenance because if you build an adhoc binding, the maintaince costs are transferred to the application code instead of the library code (which is what the library is trying to achieve: hiding the weird javascripty parts from the application code)

**Comment by cmeeren** [comment]
Hm, let's see.

Say I have the following binding type for some JS-native object:

```f#
type Mixins =
  abstract toolbar: obj with get, set

type Theme =
  abstract mixins: Mixins with get, set
  // lots of other members
```

I'd like to use `theme.mixins.toolbar` as `IStyleAttribute []`. Since `toolbar: obj` must exist in one form or another, this leaves me two options:

### 1. Use them as abstract classes instead of interfaces, and add members with other names

```f#
[<AbstractClass>]
type Mixins =
  abstract toolbar: obj with get, set
  member this.toolbarStyles =
    objectEntries this.toolbar |> unbox<IStyleAttribute []>
```

* Pros: Simple change, low maintenance, no quasi-duplicated types (which can also be confusing for users - for example, the above object may be created directly by the user, for example using `jsOptions`, when creating a custom theme)
* Cons: Bigger size, since it now compiles as a class. Also unsure if `jsOptions` etc. will create a POJO. Fable REPL [indicates it will](https://fable.io/repl/#?code=PYBwpgdgBAYghgIwDZgHQGFgCcwChSSyIobZoBSAzgJIQAuYWouuA2gDwCCCldWcAYzrokcSpQB8AXVx0AnuFjBgUALy4oURL35ConAFxQAlvSgB3Y3QAWUAOZg6AGiiVHGqAFswnhIyg2xpSoAEJqAdZBqJwsKHRQAB7hAFaUAPIgdMbAEJQAFABmAK7QeQlGMMoAlFAAtBKJ0VDstVAATFW4IUzmblioACLAnqgCOZTAJEjAdo0x3cC9jIPDo+OTaNOzCaFAA&html=DwCwLgtgNgfAUHUBTAhgE3gAm54ElgqbhgAOAtEgI4CuAlgG4C8A5AMID2AdmEj+QBUAnqSQtMAY268erXgA8wAenDQA3JJAoATgGcCTGmABm5ABwssOYLona6pMJl3aJTAEQB9TwAkA8gDKAt5KUHQARrpK2qgSYAB0pNocaDRxdNzxEHRc8QBWuu4wwEq29o5W2DZ2Dk4ubl6+gcGeoRFRMShx5GgcEInJqemZ2bkFRSVltfAlIKgYCMDhKUKSUCi6uh4opBRSPCg5SNpFcDi4aIyYdGjbu+5rG1vuO6QTSpcMM0rLaELwiBUkFgcCAA&css=BYFwtgNgNABARgewCYE8YG8BQMbAKYCWA5qAFwwCMADFQKQDc2MYAhgE5EEB25VjOCAG542AMwgIA7uWAEkSPF0YBfTJgDELAA5aMTfMTKUaDTKswA6VtwC0AYwRcQLbiL04DJEOWp1+8FjsAayI2BABXLiR7BAk2cjgIQKCVNQtEAA93GC0EAGcCEAJHchY4PNjwkDx-IRFxKRk5BSUzNMzSYDq2bIiQCFcfLSzJWWqYCoGkVMtuPJA2cLsixzzs3IKVnhhRAgy8aaYQBC1efzZDbxg+Jjhk0IiomLjyUcKapkk5EGByACYaMN-KJHCAbAUAF54Hx-IFMEFOGyiFhgAgQFDkPIsLh5cEiAiifxaFjybhEcgAVjh5gscwWSy2azgVWOPF2bHmNgQohsIBQWjw2VYHFsFy8Pio1MwQA), but that may be an implementation detail for all I know.

### 2. Have "native" bindings in addition to customized bindings

This is what I indicated in my previous comment. In addition to `Mixins`, `Theme`, we define:

```f#
type FelizMixins =
  abstract toolbar: IStyleAttribute list

type FelizTheme =
  abstract mixins: FelizMixins with get, set
  // lots of other members
```

Then we use this instead.
* Pros: We can stick to interfaces
* Cons: Every function in the JS API that accepts or returns a normal `Theme` must be wrapped in order to convert between `FelizTheme` and `Theme`.

Please ask if I'm not being clear enough.

**Comment by cmeeren** [comment]
Oh wait, it seems I can use extension members for option 1, meaning that some of the cons no longer apply.

**Comment by Zaid-Ajaj** [comment]
@cmeeren no extension methods needed, try this for option 1
```fs
[<AbstractClass; Erase>]
type Mixins =
  [<Emit("Object.entries($0.toolbar)")>]
  member this.toolbar : IStyleAttribute [] = jsNative
```

**Comment by cmeeren** [comment]
Thanks! That's even better. I'll try to use that when I can. Do you have to use `jsNative`? I.e., no F# code for those members?

**Comment by Zaid-Ajaj** [comment]
When you use `[<Emit(...)>]`, the expression on the left hand side is ignored altogether (actually if `jsNative` was to run, it throws an exception "js only" or something like that)

**Comment by cmeeren** [comment]
Sorry I was unclear; I was wondering if I had to use `[<Emit(...)>]` or if it was possible to use your suggestion also with F# code for the member. But that [doesn't seem to work](https://fable.io/repl/#?code=PYBwpgdgBAYghgIwDZgHQGFgCcwChSSyIobZoBSAzgJIQAuYWouuA2gDwCCCldWcAYzrokcSpQDcUAKL9KYAHwBdXHQCe4WMGBQAvLihREvfkKicAXFACW9KAHdrdABZQA5mDoAaKPLoGoAFswQIRGKBdrSlQAIT0oAGYWFDooAA94gCtKAHkQOmtgCEoACgAzAFdoErSrGG0ASigAWgV01E4odmaoACYG3Bime3ksVAARYEDUASLKYBIkYDd2zkHh0YmpmbmFtCWVtNjcIA&html=DwCwLgtgNgfAUHUBTAhgE3gAm54ElgqbhgAOAtEgI4CuAlgG4C8A5AMID2AdmEj+QBUAnqSQtMAY268erXgA8wAenDQA3JJAoATgGcCTGmABm5ABwssOYLona6pMJl3aJTAEQB9TwAkA8gDKAt5KUHQARrpK2qgSYAB0pNocaDRxdNzxEHRc8QBWuu4wwEq29o5W2DZ2Dk4ubl6+gcGeoRFRMShx5GgcEInJqemZ2bkFRSVltfAlIKgYCMDhKUKSUCi6uh4opBRSPCg5SNpFcDi4aIyYdGjbu+5rG1vuO6QTSpcMM0rLaELwiBUkFgcCAA&css=BYFwtgNgNABARgewCYE8YG8BQMbAKYCWA5qAFwwCMADFQKQDc2MYAhgE5EEB25VjOCAG542AMwgIA7uWAEkSPF0YBfTJgDELAA5aMTfMTKUaDTKswA6VtwC0AYwRcQLbiL04DJEOWp1+8FjsAayI2BABXLiR7BAk2cjgIQKCVNQtEAA93GC0EAGcCEAJHchY4PNjwkDx-IRFxKRk5BSUzNMzSYDq2bIiQCFcfLSzJWWqYCoGkVMtuPJA2cLsixzzs3IKVnhhRAgy8aaYQBC1efzZDbxg+Jjhk0IiomLjyUcKapkk5EGByACYaMN-KJHCAbAUAF54Hx-IFMEFOGyiFhgAgQFDkPIsLh5cEiAiifxaFjybhEcgAVjh5gscwWSy2azgVWOPF2bHmNgQohsIBQWjw2VYHFsFy8Pio1MwQA) (`ReferenceError: Foo$$get_B is not defined`).

**Comment by Zaid-Ajaj** [comment]
Works if the member is `inline`
```fs
[<AbstractClass; Erase>]
type Foo =
  abstract A: int with get, set
  member inline this.B = this?A + 5
```

**Comment by cmeeren** [comment]
Oh, all the trickery! 😵 Thanks a lot!

**Comment by Zaid-Ajaj** [comment]
Thanks! happy to help :smile:

**Comment by cmeeren** [comment]
...Though that has the same considerations as other `inline` stuff, right? That if the body is not very short, it might be better to make it non-inline? I.e. use the `Emit` trick or extension members?

**Comment by Zaid-Ajaj** [comment]
Not very short *and* used in a number of places => do not inline. if it used once or twice in your app to call a couple of functions inline, no problem at all.

**Comment by Zaid-Ajaj** [comment]
Can we close this issue @cmeeren?

**Comment by cmeeren** [comment]
Sure!
